src/ffi/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,127 @@
+/// to its address 
+#[no_mangle]
+pub extern "C" fn new_quaternion(real: f64, i: f64, j: f64, k: f64) -> *mut Quaternion {
+    Quaternion::new(real, i, j, k).to_raw_pointer()
+}
+
+/// Initialize a quaternion from a real part and a C pointer to a Vector3
+/// and retrieve the C pointer to its address 
+#[no_mangle]
+pub unsafe extern "C" fn new_quaternion_from_vector(
+    real: f64, imag_ptr: *const Vector3
+) -> *mut Quaternion {
+    null_pointer_check!(imag_ptr);
+    let imag = *imag_ptr;
+    Quaternion::new_from_vector(real, imag).to_raw_pointer()
+}
+
+/// Free memory at the address of the quaternion pointer. Outer processes
+/// that work with Quaternions via the FFI interface MUST remember 
+/// to call this function when finished with a quaternion
+#[no_mangle]
+pub unsafe extern "C" fn free_quaternion_memory(ptr: *mut Quaternion) {
+    if ptr.is_null() {
+        return;
+    }
+    Box::from_raw(ptr);
+}
+
+/// Get the components of a quaternion as a list of C doubles, the order of the
+/// components will be (real, i, j, k)
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_get_components(quat_ptr: *const Quaternion) -> *const c_double {
+    null_pointer_check!(quat_ptr);
+    let quat = *quat_ptr;
+    let components: [c_double;4] = [quat.real, quat.i, quat.j, quat.k];
+    Box::into_raw(Box::new(components)) as *const _
+}
+
+/// Set the real component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_real(quat_ptr: *mut Quaternion, real: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.real = real
+}
+
+/// Set the i component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_i(quat_ptr: *mut Quaternion, i: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.i = i
+}
+
+/// Set the j component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_j(quat_ptr: *mut Quaternion, j: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.j = j
+}
+
+/// Set the k component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_k(quat_ptr: *mut Quaternion, k: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.k = k;
+}
+
+/// Set the imaginary components of an existing quaternion stored at
+/// the address of a pointer (quat_ptr) from the components of a 3-vector
+/// (stored at vec_ptr). The convention is x -> i, y -> j, z -> k
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_imag_from_vector(quat_ptr: *mut Quaternion, vec_ptr: *const Vector3) {
+    null_pointer_check!(quat_ptr);
+    null_pointer_check!(vec_ptr);
+    let mut quat = *quat_ptr;
+    let imag = *vec_ptr;
+    quat.set_imag_from_vector(imag);
+}
+
```

Comment:
[nit] Is there a need/want for a set_quaternion function? I  noticed you have a get for all four elements, a set for each real and imaginary component, and a set for imaginary only. Wanted to point it out, I think it's fine/less error prone as is.

Reply:
yeah I'll make a set_from_components